src/serializationStrategies/XpathRangeSelector/xpath.js

```diff
@@ -4,32 +4,62 @@ const findElementChild = node => Array.prototype.find.call(node.childNodes, node
 const isHighlight = node => node && node.getAttribute && node.getAttribute(DATA_ATTR) !== null;
 const isTextHighlight = node => isHighlight(node) && !findElementChild(node);
 const isText = node => node && node.nodeType === 3;
+const isTextOrTextHighlight = node => isText(node) || isTextHighlight(node);
 const isElement = node => node && node.nodeType === 1;
-const isTextOrElement = node => isText(node) || isElement(node);
+const isElementNotHighlight = node => isElement(node) && !isHighlight(node);
 
 const IS_PATH_PART_SELF = /^\.$/;
 const IS_PATH_PART_TEXT = /^text\(\)\[(\d+)\]$/;
 const IS_PATH_PART_ELEMENT = /\*\[name\(\)='(.+)'\]\[(\d+)\]/;
 
 // kinda copied from https://developer.mozilla.org/en-US/docs/Web/XPath/Snippets#getXPathForElement
 export function getXPathForElement(targetElement, offset, reference) {
+
+  // if the range offset designates text or a text highlight we need to move the target
+  // so text offset stuff will work
+  if (isElement(targetElement) && isTextOrTextHighlight(targetElement.childNodes[offset])) {
+    targetElement = targetElement.childNodes[offset];
+    offset = 0;
+  }
+
   let xpath = '';
   let pos
     , element = targetElement.previousSibling
     , focus = targetElement;
 
   // for a text target, highlights might have broken up the text node,
   // look for preceeding nodes that need to be combined into this one
-  // and modify the range offset accordingly
+  // and modify the range offset accordingly. only have to look at one
+  // previous sibling because text nodes cannot be siblings
   if (isText(focus) && isTextHighlight(element)) {
-    while(element && (isText(element) || isTextHighlight(element))) {
+    while(isText(element) || isTextHighlight(element)) {
+      offset += element.textContent.length;
+      element = element.previousSibling;
+    }
+  // if target is text highlight, treat it like its text
+  } else if (isTextHighlight(focus) && isTextOrTextHighlight(element)) {
+    offset = 0;
+
+    while(isText(element) || isTextHighlight(element)) {
       offset += element.textContent.length;
       element = element.previousSibling;
     }
   // for element targets, highlight children might be artifically
   // inflating the range offset, fix.
   } else if (isElement(focus)) {
-    offset -= Array.prototype.filter.call(Array.prototype.slice.call(focus.childNodes, 0, offset), isHighlight).length;
+    let search = focus.childNodes[offset];
+
+    while (search) {
+      if (isTextOrTextHighlight(search)) {
+        search = search.previousSibling;
+
+        while(isTextOrTextHighlight(search)) {
+          offset--;
+          search = search.previousSibling;
+        }
+      }
+      search = search ? search.previousSibling : null;
+    }
   }
```

Comment:
These nested while loops scare me, since it seems like a N+1 waiting to happen if something fails or if the input is larger than expected.

it's also not immediately obvious to me what they're doing.  Is there anyway to tuck each of them into well tested functions who's name would make their purpose clear?

Reply:
its going backwards through the elements looking for text nodes or highlights, then when it finds one it goes backwards through any subsequent text or highlight nodes subtracting them from the offset because in the actual document once the highlights are all taken out all adjacent text/highlight nodes will actually be one text node.

I know the cognitive complexity is too high in these functions, crawling doms kind of sucks. i plan to try to refactor the logic into more manageable chunks once i have enough test coverage that i'm confident i don't break it at the same time.